### PR TITLE
[SPARK-39081][PYTHON][SQL] Implement DataFrame.resample and Series.resample

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
Implement DataFrame.resample and Series.resample

### Why are the changes needed?
To Increase pandas API coverage in PySpark

### Does this PR introduce _any_ user-facing change?
yes, new methods added

for example:
```
In [3]: 
   ...: dates = [
   ...:     datetime.datetime(2011, 12, 31),
   ...:     datetime.datetime(2012, 1, 2),
   ...:     pd.NaT,
   ...:     datetime.datetime(2013, 5, 3),
   ...:     datetime.datetime(2022, 5, 3),
   ...: ]
   ...: pdf = pd.DataFrame(np.ones(len(dates)), index=pd.DatetimeIndex(dates), columns=['A'])
   ...: psdf = ps.from_pandas(pdf)
   ...: psdf.resample('3Y').sum().sort_index()
   ...: 
Out[3]: 
              A
2011-12-31  1.0
2014-12-31  2.0
2017-12-31  0.0
2020-12-31  0.0
2023-12-31  1.0



```

### How was this patch tested?
added UT

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
This is amazing!

**Comment by HyukjinKwon** [comment]
cc @itholic @xinrong-databricks mind taking a look when you find some time please 🙏

File: python/pyspark/pandas/resample.py [hunk 1]
```diff
@@ -0,0 +1,528 @@
...
+        if len(parsed) != 1:
+            raise ValueError("Unsupported freq {}".format(rule))
+
+        offset_str, unit_str = parsed[0]
+        self._freq_offset = 1
+        if offset_str != "":
+            freq_offset = int(offset_str)
+            if not freq_offset > 0:
+                raise ValueError("invalid rule: '{}'".format(rule))
+            self._freq_offset = freq_offset
```
Comment (xinrong-databricks) [inline python/pyspark/pandas/resample.py:111]: Would you think we may leverage `from pandas._libs.tslibs import to_offset` here?

```py
>>> to_offset("3MIN").freqstr
'3T'
>>> to_offset("3MIN").n
3
>>> to_offset("3MIN").rule_code
'T'
>>> to_offset("3MI")
...
ValueError: Invalid frequency: MI
```

File: python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst [hunk 2]
```diff
@@ -362,7 +362,9 @@ Supported DataFrame APIs
 +--------------------------------------------+-------------+--------------------------------------+
 | :func:`replace`                            | P           | ``regex``, ``method``                |
 +--------------------------------------------+-------------+--------------------------------------+
-| resample                                   | N           |                                      |
+| resample                                   | P           |``axis``, ``convention``, ``kind``    |
```
Comment (itholic) [inline python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst:365]: Can we add the link??

```
:func:`resample`
```

should work

File: python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst [hunk 3]
```diff
@@ -1016,7 +1018,9 @@ Supported Series APIs
 | :func:`replace`                 | P                 | ``inplace``, ``limit``, ``regex``,        |
 |                                 |                   | ``method``                                |
 +---------------------------------+-------------------+-------------------------------------------+
-| resample                        | N                 |                                           |
+| resample                        | P                 |``axis``, ``convention``, ``kind``         |
```
Comment (itholic) [inline python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst:1021]: ditto!

File: python/pyspark/pandas/resample.py (hunk 1 above)
Comment (zhengruifeng) [reply python/pyspark/pandas/resample.py:111]: good idea. I don't know this utils before, will update, thanks!

File: python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst (hunk 2 above)
Comment (zhengruifeng) [reply python/docs/source/user_guide/pandas_on_spark/supported_pandas_api.rst:365]: good catch! will update

**Review by xinrong-meng — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]
LGTM2. Will take another look tomorrow and get this in.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**